Add support for WebGL2 framebuffer attachments

mmatyas opened the pull request:
Adds an initial implementation for the framebuffer attachments introduced with WebGL2 and the related enums and constrains checks.

<!-- Please describe your changes on the following line: -->
Depends on #25765.

cc @jdm @zakorgy 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webgltexture.rs, components/script/dom/webglframebuffer.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webglrenderbuffer.rs
 * @jgraham: tests/wpt/webgl/meta/conformance2/rendering/framebuffer-completeness-unaffected.html.ini, tests/wpt/webgl/meta/conformance2/rendering/framebuffer-texture-level1.html.ini, tests/wpt/webgl/meta/conformance2/state/gl-object-get-calls.html.ini, tests/wpt/webgl/meta/conformance2/renderbuffers/framebuffer-test.html.ini, tests/wpt/webgl/meta/conformance2/renderbuffers/framebuffer-object-attachment.html.ini and 2 more
 * @kichjang: components/script/dom/webgltexture.rs, components/script/dom/webglframebuffer.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webglrenderbuffer.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

Manishearth commented:
r? @nox  or @jdm 

I'm less familiar with a lot of the webgl code

nox commented:
The implementation looks good, but could you please amend the three commits to recreate the expectations from scratch? Your fixes look like they only added failures because the test cases in the Khronos test suites are weirdly numbered programatically, and every time we support new APIs their numbers changes and most of the previous expectations become garbage.

jdm commented:
Sorry, left review comments yesterday and forgot to submit them.

jdm commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -436,7 +496,10 @@ impl WebGLFramebuffer {
     ) -> Option<&DomRefCell<Option<WebGLFramebufferAttachment>>> {
         match attachment {
-            constants::COLOR_ATTACHMENT0 => Some(&self.color),
+            constants::COLOR_ATTACHMENT0..=constants::COLOR_ATTACHMENT15 => {
```
This changes the behaviour of passing COLOR_ATTACHMENTN in webgl1 contexts, as far as I can tell.

mmatyas replied:
In WebGL1, the number of color attachments is hardcoded to 1, so the color vector will also have only one element, and asking for anything other that COLOR_ATTACHMENT0 will return None like before.

jdm commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -84,15 +89,15 @@ pub enum WebGLFramebufferAttachmentRoot {
     // The attachment points for textures and renderbuffers on this
     // FBO.
-    color: DomRefCell<Option<WebGLFramebufferAttachment>>,
+    colors: Vec<DomRefCell<Option<WebGLFramebufferAttachment>>>,
```
On second thought, there's enough code that benefits from the current type that I don't think it's worth switching.

mmatyas committed:
Add support for WebGL2 framebuffer attachments

Adds an initial implementation for the framebuffer attachments
introduced with WebGL2 and the related enums and constrains checks.

mmatyas commented on components/script/dom/webglrenderbuffer.rs:
```diff
@@ -143,25 +142,29 @@ impl WebGLRenderbuffer {
+            constants::DEPTH_COMPONENT24 |
+            constants::DEPTH_COMPONENT32F |
+            constants::DEPTH24_STENCIL8 |
+            constants::DEPTH32F_STENCIL8 |
```
Hm you might be right, I've added an additional check for this case.

mmatyas commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -213,91 +219,138 @@ impl WebGLFramebuffer {
         self.size.get()
     }
 
+    fn check_attachment_constrains(
```
Indeed, thanks!

mmatyas commented:
@nox Ah, I might have accidentally left some of such tests in the commits, will fix it.

mmatyas commented:
Also rebased to master, now that the earlier patches have landed.

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1582203730577.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 974b0f3 has been approved by `jdm`

<!-- @bors-servo r=jdm 974b0f349fa0a02993ed7fa246151cdfc743e957 -->

bors-servo commented:
:hourglass: Testing commit 974b0f349fa0a02993ed7fa246151cdfc743e957 with merge e46ed2b996cba10d1cf2cb0fbf7ee2d9c0ce288c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/cEYIsMnIRgi9SMfl--g31g)

jdm commented:
```
  ▶ Unexpected subtest result in /_webgl/conformance2/state/gl-object-get-calls.html:
  │ FAIL [expected PASS] WebGL test #182: successfullyParsed should be true (of type boolean). Was undefined (of type undefined).
  │   → assert_true: successfullyParsed should be true (of type boolean). Was undefined (of type undefined). expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1867:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:7
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:5
  │ shouldBe@http://web-platform.test:8000/_webgl/js/js-test-pre.js:406:9
  │ shouldBeTrue@http://web-platform.test:8000/_webgl/js/js-test-pre.js:432:29
  └ @http://web-platform.test:8000/_webgl/js/js-test-post.js:24:1
```

mmatyas commented:
Hm weird, it didn't happen to me, but I've regenerated the test just in case.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit b41805a has been approved by `jdm`

<!-- @bors-servo r=jdm b41805a9206f940737c2644f73d975ae35e38f99 -->

bors-servo commented:
:hourglass: Testing commit b41805a9206f940737c2644f73d975ae35e38f99 with merge 4813ec2fce6420c98425b142b2b591b32fcd862b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/VJPONh0oSaCez3UVSLXFgQ)

jdm commented:
@bors-servo retry
* #24726

bors-servo commented:
:hourglass: Testing commit b41805a9206f940737c2644f73d975ae35e38f99 with merge 05b67263aac7f4c74e36bffa1126cbf4b8274325...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Zk5FFCrsTMWQAbxqy5vJwQ)

jdm commented:
@bors-servo retry
* #24726

bors-servo commented:
:hourglass: Testing commit b41805a9206f940737c2644f73d975ae35e38f99 with merge 947a60224a7e6198429e80eba3abd547d4a41891...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/dAnZp8TGTwmrxJmQMmbL0g)

jdm commented:
@bors-servo retry
* #24726

bors-servo commented:
:hourglass: Testing commit b41805a9206f940737c2644f73d975ae35e38f99 with merge afb1faf80f3efbdd2611adad74a125b37df25f61...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/RjZrQu47RDOnVwQVkl-ZtQ)
Approved by: jdm
Pushing afb1faf80f3efbdd2611adad74a125b37df25f61 to master...
